### ISSUE TITLE
Deploying multiple problems with same name leads to bad state

### ISSUE THREAD

**Issue by dmartin** [picoCTF/picoCTF#268]
Bug report:
> If two different bundles contain problems with the same sanitized name, neither will work.
They show up in the problem list in Management on the web and when shell_manager status is called, but they don't appear in the actual problem list either on the web or through the api
Trying to query the problem id through the API causes this exception:
```
Traceback (most recent call last):
  File "/picoCTF/picoCTF-web/api/annotations.py", line 62, in wrapper
    web_result = f(*args, **kwds)
  File "/picoCTF/picoCTF-web/api/annotations.py", line 94, in wrapper
    return f(*args, **kwds)
  File "/picoCTF/picoCTF-web/api/annotations.py", line 171, in wrapper
    return f(*args, **kwds)
  File "/picoCTF/picoCTF-web/api/annotations.py", line 194, in wrapper
    return f(*args, **kwds)
  File "/picoCTF/picoCTF-web/api/routes/problem.py", line 96, in get_single_problem_hook
    problem_info = api.problem.get_problem(pid, tid=api.user.get_user()['tid'])
  File "/picoCTF/picoCTF-web/api/cache.py", line 172, in wrapper
    function_result = f(*args, **kwargs)
  File "/picoCTF/picoCTF-web/api/problem.py", line 710, in get_problem
    raise InternalException("You cannot get this problem")
api.common.InternalException: You cannot get this problem
```

- Replicate and understand this behavior
- Possible short-term solution: check existing problem names before deployment and throw an exception before deploying one with the same name
- Longer term, can we refactor to avoid using the sanitized problem name as a unique identifier?

**Comment by syreal17** [comment]
After discussing with Max who has changed some Python files for more careful deployment of problems:
* One of the biggest issue is that _installing_ problems happens through Linux package manager (apt-get/dpkg)
   * Right now, problems with the same `deb` package name (which is different than filename and is referred to by "sanitized name" in shell manager code) immediately overwrite each other with no warning or even report of overwriting
   * This is intended behavior for APT for packages of high priority.
   * We could instead have shell manager assign a low priority to problem `deb` packages so that they never overwrite each other... **but** this might really trading pros and cons
      * "A package with a priority ranging between 0 and 100 will only be installed if no other version of the package is already installed." from: https://kali.training/topic/advanced-apt-configuration-and-usage/

**Comment by syreal17** [comment]
@hiliang-cmu and I discussed a better naming convention for problems moving forward, esp. as we anticipate our platform persisting for more than one competition:

```
sanitized_problem_name_[trimmed_hash_material]
```

Where `[trimmed_hash_material]` is the last 7 digits of the result of hashing year, author, organization, etc.

As an example, what is now: `buffer_overflow_1` would be `buffer_overflow_1_1784gdc`

**Comment by syreal17** [comment]
Perhaps the most practical step, would be applying my stated naming convention above to the example problems that come standard in our platform.

**Comment by hiliang-cmu** [comment]
Also hashing `year`, `organization` or somesuch will be needed in web api too for problemid generation:

https://github.com/picoCTF/picoCTF/blob/2072ac5c7d036eeb4af34a079d732298f3a17d00/picoCTF-web/api/problem.py#L101

If two `buffer_overflow_1` exist by same author exist, that'll mess up the web platform too

**Comment by syreal17** [comment]
@hiliang-cmu @dmartin please comment if you like any tweaks to the plan
@MongYahHsieh short summary of our plan for this issue:
We're trying to ensure that problem names are extraordinarily unlikely to be the same. Right now, it's not that surprising that we've already had two problems called "buffer-overflow-1" when our naming scheme is just `sanitized_problem_name`. Let's implement a new naming scheme, and update the default loaded problems to use it so we avoid name collisions in the future.

design/implementation high points:
1. Adopt the following naming convention:
   * `sanitized_problem_name_[trimmed_hash_material]` where:
      * `[trimmed_hash_material]` is the last 7 digits of the result of `hashing_function(prob.name, prob.year, prob.author, prob.organization)`
         * `hashing_function` might be best if it is a fast hashing function like MD5. For this application, we do not need a cryptographically secure hashing function, but we do not want any collisions either. **Please do a little research along these lines and propose a hashing function that you want to use.**
         * To explain, we take just the last 7 digits of the hash to avoid annoyingly long problem names (hashes can be very long), but to get repeatable randomness as well. As a sidenote, Git does something similar with commit hashes.
2. Apply naming convention to example problems loaded with picoCTF platform by default
   1. Apply the naming convention manually to the example problems to get the hang of it
   2. Apply the naming convention automatically with a (Python) script to the example problems so that you know almost exactly what to do for shell manager...
3. Apply naming convention to shell manager so that packaging a problem uses our new naming convention
4. _(This last step is not high priority)_ Investigate package priorities to help prevent deployers from overwriting unintended problems in the unlikely scenario that there are duplicate names (after implementing the new naming convention, this will have more to do with our naming convention being manually superseded...)

**Comment by hiliang-cmu** [comment]
> [trimmed_hash_material] is the last 7 digits of the result of hashing_function(prob.name, prob.year, prob.author, prob.organization)

As mentioned in my post above, problem id (`pid`) is currently generated on the web server but we might as well have the shell_manager generate it when loading problems. Just use this hash scheme, save it as part of problem metadata (and use those 7 digits as part of packaged problem name), and pass it along to webserver when it loads problems. 

The

**Comment by MongYahHsieh** [comment]
I've added hashing on the shell side: 3272ce9067716b419414fffca286b609f08c8741
All problems folders are now generated with a seven-character MD5 hash identical to the first 7 characters of the pid the web server generates.

The function generating the hash is `get_pid_hash()` in `/picoCTF-shell/shell_manager/util.py`, which is where the fields used to generate the hash is defined.

I've also modified the names of some variables and the documentation of several functions to better show what they actually are/do now.
